File: LinkedList.c
#include <stdio.h>
#include <stdlib.h>

typedef struct ListNode
{
    int value;
    struct ListNode *next;
} ListNode;

ListNode *CreateNode(int value)
{
    ListNode *ln = malloc(sizeof(ListNode));
    if (ln == NULL)
    {
        printf("Failed to allocate memory\n");
        exit(-1);
    }
    else
    {
        ln->value = value;
        ln->next = NULL;
    }
    return ln;
}

void add(ListNode **head, int value)
{
    ListNode *newNode = CreateNode(value);
    if (*head == NULL)
    {
        *head = newNode;
    }
    else
    {
        ListNode *current = *head;
        while (current->next != NULL)
        {
            current = current->next;
        }
        current->next = newNode;
    }
}

void prepend(ListNode **head, int value)
{
    ListNode *newNode = CreateNode(value);
    newNode->next = *head;
    *head = newNode;
}
void delete(ListNode **head, int value)
{
    ListNode *current = *head, *prev = NULL;
    if (current != NULL && current->value == value)
    {
        *head = current->next;
        free(current);
        return;
    }

    while (current != NULL && current->value != value)
    {
        prev = current;
        current = current->next;
    }

    if (current == NULL)
        return;

    prev->next = current->next;
    free(current);
}

void fibonacci(ListNode **head, int value) {
    if (value <= 0) {
        printf("The number of Fibonacci terms must be positive.\n");
        return;
    }

    ListNode *temp;
    while (*head != NULL) {
        temp = *head;
        *head = (*head)->next;
        free(temp);
    }

    if (value >= 1) {
        *head = CreateNode(0); 
        if (value == 1) return;
    }

    ListNode *current = CreateNode(1);
    (*head)->next = current;

    int first = 0, second = 1, sum;
    for (int i = 2; i < value; i++) {
        sum = first + second;
        current->next = CreateNode(sum);
        current = current->next;
        first = second;
        second = sum;
    }
}


int get(ListNode *head, int index)
{
    int count = 0;
    ListNode *current = head;
    while (current != NULL)
    {
        if (count == index)
        {
            return current->value;
        }
        current = current->next;
        count++;
    }
    printf("Index out of bounds\n");
    return -1; 
}

void set(ListNode** head, int index, int newValue) {
    int count = 0;
    ListNode* current = *head;
    while (current != NULL) {
        if (count == index) {
            current->value = newValue;
            return;
        }
        current = current->next;
        count++;
    }
    printf("Index %d out of bounds\n", index);
}


void insertAt(ListNode **head, int index, int value)
{
    ListNode *newNode = CreateNode(value);
    ListNode *current = *head;
    ListNode *prev = NULL;
    int i = 0;

    if (index == 0)
    {
        prepend(head, value);
        return;
    }

    while (i < index && current != NULL)
    {
        prev = current;
        current = current->next;
        i++;
    }

    if (i != index)
    {
        printf("Index out of bounds\n");
        free(newNode);
        return;
    }

    prev->next = newNode;
    newNode->next = current;
}

void printList(ListNode *head)
{
    ListNode *current = head;
    while (current != NULL)
    {
        printf("%d -> ", current->value);
        current = current->next;
    }
    printf("NULL\n");
}

int main(int argc, const char *argv[]) {
    ListNode *head = NULL;

    // Append elements
    add(&head, 1);
    add(&head, 2);
    add(&head, 3);
    printf("After adding 1, 2, 3: ");
    printList(head);

    // Test get
    int getIndex = 1;
    int getValue = get(head, getIndex);
    printf("Value at index %d: %d\n", getIndex, getValue);

    // Test set
    set(&head, 1, 20);
    printf("List after setting index 1 to 20: ");
    printList(head);

    // Test insert at index
    insertAt(&head, 2, 25);
    printf("After inserting 25 at index 2: ");
    printList(head);

    // Test delete
    delete(&head, 1);
    printf("After deleting value 1: ");
    printList(head);

    // Test prepend
    prepend(&head, 0);
    printf("After prepending 0: ");
    printList(head);

    // Generate and test Fibonacci sequence
    fibonacci(&head, 10);
    printf("Fibonacci sequence with 10 terms: ");
    printList(head);

    // Cleanup: Free all nodes
    ListNode *temp;
    while (head != NULL) {
        temp = head;
        head = head->next;
        free(temp);
    }

    return 0;
}

File: LinkedList.java
import java.util.List;

public class LinkedList {
    private static class ListNode {
        int value;
        ListNode next;

        ListNode(int value) {
            this.value = value;
            this.next = null;
        }
    }

    private ListNode head;

    public LinkedList() {
        this.head = null;
    }

    void add(int value) {
        ListNode newNode = new ListNode(value);

        if (this.head == null) {
            this.head = newNode;
            return;
        }

        ListNode current = this.head;

        while (current.next != null) {
            current = current.next;
        }
        current.next = newNode;
    }

    public void fibonacci(int value) {
        if (value <= 0) {
            System.out.println("The number of Fibonacci terms must be positive.");
            return;
        }
    
        this.head = new ListNode(0); 
        if (value == 1) return; 
    
        ListNode current = new ListNode(1);
        this.head.next = current;
    
        int first = 0, second = 1; 
    
        for (int i = 2; i < value; i++) {
            int nextFib = first + second;
            current.next = new ListNode(nextFib); 
            current = current.next; 
            first = second; 
            second = nextFib; 
        }
    }
    

    void prepend(int value) {
        ListNode newNode = new ListNode(value);
        newNode.next = this.head;
        this.head = newNode;
    }

    void print() {
        ListNode current = this.head;
        while (current != null) {
            System.out.print(current.value + " -> ");
            current = current.next;
        }
        System.out.println("null");
    }

    void delete(int value) {
        while (this.head != null && this.head.value == value) {
            this.head = this.head.next;
        }

        if (this.head == null) {
            return;
        }

        ListNode current = this.head;
        while (current.next != null) {
            if (current.next.value == value) {
                current.next = current.next.next;
            } else {
                current = current.next;
            }
        }
    }

    int size() {
        int count = 0;
        ListNode current = this.head;
        while (current != null) {
            count++;
            current = current.next;
        }
        return count;
    }

    public int get(int index) {
        if (index < 0 || index >= this.size()) {
            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + this.size());
        }
        ListNode current = this.head;
        for (int i = 0; i < index; i++) {
            current = current.next;
        }
        return current.value;
    }

    public void set(int index, int value) {
        if (index < 0 || index >= this.size()) {
            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + this.size());
        }
        ListNode current = this.head;
        for (int i = 0; i < index; i++) {
            current = current.next;
        }
        current.value = value;
    }

    void insertAt(int index, int value) {
        if (index < 0) {
            throw new IndexOutOfBoundsException("Index cannot be negative.");
        }

        ListNode newNode = new ListNode(value);
        if (index == 0) {
            newNode.next = this.head;
            this.head = newNode;
            return;
        }

        ListNode current = this.head;
        for (int i = 0; current != null && i < index - 1; i++) {
            current = current.next;
        }

        if (current == null) {
            throw new IndexOutOfBoundsException("Index is out of bounds.");
        }

        newNode.next = current.next;
        current.next = newNode;
    }
    public static void main(String[] args) {
        LinkedList list = new LinkedList();
    
        // add Test
        list.add(10);
        list.add(20);
        list.add(30);
        System.out.print("add Test: ");
        list.print();
    
        // Prepend Test
        list.prepend(5);
        System.out.print("Prepend Test: ");
        list.print();
    
        // Size Test
        System.out.println("Size Test: " + list.size());
    
        // Get Test
        try {
            System.out.println("Get Test (index 1): " + list.get(1));
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }
    
        // Set Test
        try {
            list.set(1, 15);
            System.out.print("Set Test (set index 1 to 15): ");
            list.print();
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }
    
        // Insert At Test
        try {
            list.insertAt(2, 25);
            System.out.print("Insert At Test (insert 25 at index 2): ");
            list.print();
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }
    
        // Delete Test
        list.delete(30);
        System.out.print("Delete Test (delete 30): ");
        list.print();
    
        // Fibonacci Test
        LinkedList fibList = new LinkedList();
        fibList.fibonacci(8);
        System.out.print("Fibonacci Test: ");
        fibList.print();
    }
    
}
